pyclam/search.py

```diff
@@ -0,0 +1,196 @@
+
+        :param query: point around which to look.
+        :param k: number of closest neighbors to look for.
+        :param max_depth: optional maximum search depth.
+        :return: dictionary of index-of-neighbor -> distance-to-neighbor.
+        """
+        if k < 1:
+            raise ValueError(f'k must be a positive integer. Got {k}')
+
+        radius: float = (k / self.root.cardinality) * self.root.radius
+        assert radius > 0, f'expected a positive value for radius. Got {radius:.2e} instead.'
+
+        hits: Results = self.rnn(query, radius, max_depth=max_depth)
+        while len(hits) == 0:  # make sure to have non-zero hits
+            radius *= 2
+            hits = self.rnn(query, radius, max_depth=max_depth)
+
+        while len(hits) < k:  # make sure to have more at least k hits
+            # calculate lfd in ball around query
+            if len(hits) == 1:
+                lfd = 1
+            else:
+                half_count = len([point for point, distance in hits.items() if distance <= radius / 2])
+                lfd = 1 if half_count == 0 else np.log2(len(hits) / half_count)
+
+            # increase radius as guided by lfd
+            factor = (k / len(hits)) ** (1 / lfd)
+            assert factor > 1, f'expected factor to be greater than 1. Got {factor:.2e} instead.'
+            radius *= factor
+
+            # rerun rnn search
+            hits = self.rnn(query, radius, max_depth=max_depth)
+
+        # sort hits in non-decreasing order of distance to query
+        results: List[Tuple[float, int]] = list(sorted([(distance, point) for point, distance in hits.items()]))
+        return {point: distance for distance, point in results[:k]}
+
+    def knn_points(self, query: Datum, k: int, *, max_depth: Optional[int] = None) -> Data:
+        """ Same as knn, but returns the raw points rather than a dictionary of indices and distances. """
+        results = [(d, p) for p, d in self.knn(query, k, max_depth=max_depth).items()]
+        return self.data[[p for _, p in sorted(results)]]
+
+    def _check_shape(self, points: np.ndarray) -> None:
+        if len(points.shape) != len(self.data.shape):
+            raise ValueError(f'wrong number of dimensions in shape. Expected {len(self.data.shape)}, but got {len(points.shape)}')
+        elif points.shape[1:] != self.data.shape[1:]:
+            raise ValueError(f'wrong dimensionality of points. Expected {self.data.shape[1:]}, but got {points.shape[1:]}')
+        else:
+            return
+
+    def _parse_query(self, query: Datum) -> np.ndarray:
+        if isinstance(query, int):
+            query: np.ndarray = np.asarray([self.data[query]])
+        query = np.expand_dims(query, 0)
+        self._check_shape(query)
+        return query
+
+    def tree_search(self, query: Datum, radius: float, start: Cluster, *, max_depth: Optional[int] = None) -> ClusterResults:
+        """ Performs tree-search for the query, starting at the given cluster.
```

Comment:
I think a method this important should have a bit more documentation, especially describing what it does / what it returns.

Reply:
done